### PR TITLE
Add get_item_state() function

### PR DESCRIPTION
**Pull request by politas**
Adds a new get_item_state() to provide a migration path from deprecated GetItemValue

### CODE REVIEW

**Comment by CrazyIvan359** [comment]
This is just a copy of the original function with a snake case name, nothing has really changed. What I meant in our (now lost) discussion on the forum was something along the lines of this:

```python
def get_item_value(item_or_item_name, return_type=None, default=None, unit=None):
    """

    """
    type_methods = {int: "intValue", float: "floatValue", str: "toString"}
    item = validate_item(item_or_item_name) # type: t.Union[GenericItem, None]
    if item:
        state = item.getState() # type: State
        if isinstance(state, QuantityType) and unit is not None:
            try:
                state = state.toUnit(unit)
            except:
                return default
        if not isinstance(state, UnDefType):
            if return_type is None:
                return state
            elif return_type in type_methods:
                if hasattr(state, type_methods[return_type]):
                    state = getattr(state, type_methods[return_type])()
            try: # attempt to cast
                return return_type(state)
            except:
                pass
    return default
```

This gives us a much clearer result from this function. If you only give it an item you will get the state back as-is, or the default if the item doesn't exist or the state is `NULL` or `UNDEF`. If you give it a return type it will attempt to convert the state to that type, failing that it returns the default. If the state is a `QuantityType` and you specify a unit, it will convert the measurement to that unit before attempting any conversions.

I don't use this function (yet) so I am theorizing about useability here. What are your thoughts on the above?

@rkoshak What do you think of the above Rich?

**Comment by politas** [comment]
This looks much better, if I'm understanding how it works. Should we ensure that if a type and a default are given, they match up? Something like this near the start?

```python
if return_type is not None and default is not None:
    if not isinstance(default, return_type):
        LOG.warn(u"Default value '{}' is not of type `{}`.".format(str(default), str(return_type)))
        return None
```

**Comment by politas** [comment]
The only issue I have is that with the old `getItemValue()` you are guaranteed to get a valid return item of the same type that you give as default, where this will not reliably return an item of return_type unless a default is also given.

**Comment by CrazyIvan359** [comment]
> Should we ensure that if a type and a default are given, they match up?

I intentionally didn't. This is a pattern used elsewhere, even in Python. The default is provided by you, therefore you know what it is, and if you didn't provide one the default is `None` (Pythonic again).

**Comment by politas** [comment]
Ok. So if a `return_type` is given, it will try to return the item's value as that return type, but if that fails, it will then return `default` (which will be None if no default is supplied), is that correct?

**Comment by politas** [comment]
How is this as a defblock:
```python
    """
    Returns the Item's value if the Item exists and is initialized. Returned item
    type will be dependent on the item's type. If return_type is supplied, the Item 
    will be returned as that type if possible, otherwise the default will be returned.
    If no default is supplied, a non-existent/uninitialized Item or an Item that cannot 
    be converted to return_type will return None 

    Args:
        item_or_item_name (Item or str): name of the Item
        return_type (Python data type): item type to be returned. If the Item cannot be 
            converted to return_type, default will be returned 
        default (int, float, ON, OFF, OPEN, CLOSED, str, DateTime): the default
            value to be returned if the Item is non-existent/unitilialized or cannot
            be converted to return_type
        unit (QuantityType): Convert value of Item to specified units 

    Returns:
        int, float, ON, OFF, OPEN, CLOSED, str, DateTime, return_type or None: the state 
            of the Item as a suitable type, or converted to return_type if given; or the 
            default value if the Item's state cannot be converted or is NULL or UNDEF; 
            otherise None
    """
```
I haven't found any information on state.toUnit() yet, so I'm only guessing what it does.

**Comment by politas** [comment]
I'm testing this out and having some trouble with datetimes. If I set return_type to `ZonedDateTime`, it fails to cast a DateTime Item successfully. With no return_type specified, I get an `org.openhab.core.library.types.DateTimeType`, which it seems I can use comparisons on, at least.

I can successfully convert with `core.date.to_java_zoneddatetime`, but  this `type_methods` code doesn't accept it.

**Comment by politas** [comment]
Ah, I see it! Need to add the supported datetime types to that type_methods dictionary!

**Comment by politas** [comment]
Hmmmm. I've expanded type_methods to include all the datetime types:
```python
type_methods = {
    int: "intValue",
    float: "floatValue",
    str: "toString",
    datetime: "to_python_datetime",
    ZonedDateTime: "to_java_zoneddatetime",
    JodaDateTime: "to_joda_datetime",
    JavaCalendar: "to_java_calendar"
}
```
also tried with "core.date.to_xxx"

But none of those are successfully being cast. 
```python
try
    return return_type(state)
```
Is failing every time, so I'm getting None returned  (no default set)

**Comment by CrazyIvan359** [comment]
> Ok. So if a return_type is given, it will try to return the item's value as that return type, but if that fails, it will then return default (which will be None if no default is supplied), is that correct?

Yes, exactly. 

> How is this as a defblock

Not bad, I skipped writing one since I didn't have the time, I'll clean this up a little. For clarity, it's called a `docstring` in Python.

> I haven't found any information on state.toUnit() yet, so I'm only guessing what it does.

Notice where it is used? It is a method of `QuantityType` to convert to the requested units. 

`type_methods` is meant to be used as a way to call Java conversion methods available on `State` objects. Doing these date conversions is too specific, they should be left to the user code to do. Also a `DateTimeType` can be used like a Joda DateTime object in OH2, and a Java ZonedDateTime object in OH3, no conversions necessary.

**Comment by politas** [comment]
> For clarity, it's called a `docstring` in Python.

Yeah, I knew I had the terminology wrong. 

>Notice where it is used? It is a method of QuantityType to convert to the requested units.

Ok, added a link to the relevant doco and tweaked the docstring wording. 

Oh, and I added an extra `get_item_value_or_default()` function that replaces getItemValue without any further rewrite.

File: Core/automation/lib/python/core/utils.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 __all__ = [
     "validate_channel_uid",
     "validate_uid",
+    "get_item_value",
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py]: Move this to the end of list, and the function to the end of the file.

File: Core/automation/lib/python/core/utils.py [hunk 2]
```diff
@@ -33,13 +35,15 @@
 try:
     from org.joda.time import DateTime as JodaDateTime
 except:
+    # OH3 does not have Joda Time
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py]: Don't modify other parts of the code, muddies the water if a PR adding a function does random other things too.

File: Core/automation/lib/python/core/utils.py [hunk 3]
```diff
@@ -19,6 +20,7 @@
 
 import re
 import uuid
+from datetime import datetime
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py:23]: Remove unused imports

File: Core/automation/lib/python/core/utils.py [hunk 4]
```diff
@@ -33,13 +35,15 @@
 try:
     from org.joda.time import DateTime as JodaDateTime
 except:
+    # OH3 does not have Joda Time
     JodaDateTime = None
 
 from java.time import ZonedDateTime
+from java.util import Calendar as JavaCalendar
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py:40]: Remove unused imports

File: Core/automation/lib/python/core/utils.py [hunk 5]
```diff
@@ -118,6 +122,61 @@ def validate_uid(uid):
...
+                pass
+    else:
+        LOG.warn(u"Item does not exist or is uninitialized, returning default value")
+    return default
+
+def get_item_value_or_default(item_or_item_name, default_value, return_type=None, unit=None):
+    """
+    See get_item_or_default, this is a drop-in replacement for old getItemValue
+    """
+    return get_item_value(item_or_item_name, return_type=return_type, default=default_value, unit=unit)
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py]: I disagree with adding this, the new function does this out of the box. Also the docstring here is a lie, this function *cannot* be used as a drop-in replacement because it doesn't return the same result given the same inputs.

I do agree about the order of arguments, thinking about it a bit more it seems like a default would be provided more often that a return type. On that thought, what about `get_item_state` as a function name? Gives a better idea of what it does.

File: Core/automation/lib/python/core/utils.py [hunk 6]
```diff
@@ -118,6 +122,61 @@ def validate_uid(uid):
...
+    
+    type_methods = {int: "intValue", float: "floatValue", str: "toString"}
+    item = validate_item(item_or_item_name) # type: t.Union[GenericItem, None]
+    if item:
+        state = item.getState() # type: State
+        if isinstance(state, QuantityType) and unit is not None:
+            try:
+                state = state.toUnit(unit)
+            except:
+                LOG.warn(u"Item '{}' cannot be converted to '{}' units, returning default value".format(item.name, unit))
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py]: Good idea!

Let's add *state* to the message and shorten it to just describe what went wrong, returning the default is implied by the docstring already.
```python
LOG.warn(u"Item '{}' state [{}] cannot be converted to '{}' units".format(item.name, item.state, unit))
```

File: Core/automation/lib/python/core/utils.py [hunk 7]
```diff
@@ -118,6 +122,61 @@ def validate_uid(uid):
...
+            value to be returned if the Item is non-existent/uninitialized or cannot
+            be converted to return_type
+        unit (QuantityType): Convert value of Item to specified units
+            https://www.openhab.org/docs/concepts/units-of-measurement.html
+
+    Returns:
+        int, float, ON, OFF, OPEN, CLOSED, str, DateTime, or None: the state of the Item
+            as a suitable type, or converted to return_type if given; or the default value 
+            if the Item's state cannot be converted or is NULL or UNDEF; otherwise None
+    """
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py]: How about this:

```python
    """
    Get's an Item's state, optionally converting it to a specified type and/or unit for Quantity Types.

    Args:
        item_or_item_name (GenericItem or str): Item or name of the Item
        return_type (Any): Python or Java data type to attempt to return the Item's state as, if not
            provided or `None` no conversion will be done. If the conversion fails the default will
            be returned.
        default (Any): value to return if Item does not exist, if Item's state is ``NULL`` or ``UNDEF``
            or if attempts to convert Item's state fail.
        unit (str): if Item's state is a QuantityType, attempt to convert to this unit. This is done
            before any conversion attempts if ``return_type`` is provided.

    Returns:
        Item's state if Item exists and it's state is not ``NULL`` or ``UNDEF`` if no ``return_type``
        or ``unit`` is provided, otherwise ``default`` will be returned.
        If ``return_type`` is provided the Item's state will be returned as that type if possible,
        otherwise ``default`` will be returned.
        If ``unit`` is provided and the Item's state is a QuantityType it will be converted to those
        units before attempting to convert to ``return_type``, if the unit conversion fails ``defualt``
        will be returned.
    """
```

You will need to adjust the above for line length, there is no ruler in this comment box.
Don't forget to switch the order of the arguments if you change them in the function signature.

**Review by CrazyIvan359 — CHANGES_REQUESTED** [review]
On the subject of converting dates, what about this:

```python
conv_methods = {
    ZonedDateTime: to_java_zoneddatetime
    # and etc
}
```

```python
elif return_type in conv_methods:
    return_type = conv_methods[return_type]
```

That will change `return_type` from being the actual type into the function that will convert the state to that type so the call in the try block still works.

File: Core/automation/lib/python/core/utils.py (hunk 5 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: Yes, I'm discovering this as I attempt to use it as such.

I was trying to do similar to 
```python
def send_command_if_different(item_or_item_name, new_value, floatPrecision=None):
    """
    See postUpdateCheckFirst
    """
    return postUpdateCheckFirst(item_or_item_name, new_value, sendACommand=True, floatPrecision=floatPrecision)
```

File: Core/automation/lib/python/core/utils.py (hunk 7 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: I really like the idea of the URL to the explanatory doco for QuantityType, because It's quite difficult to find the relevant information, but I guess it's not reliable long-term.

File: Core/automation/lib/python/core/utils.py (hunk 6 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: Cool, I'm shortening the failed conversion warning similarly.

```python
LOG.warn(u"Item '{}' type '{}', cannot be converted to '{}'".format(item.name, type(state), return_type)
```

File: Core/automation/lib/python/core/utils.py (hunk 7 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: Much clearer! I think I'd forgotten the proper docstring format.

File: Core/automation/lib/python/core/utils.py (hunk 1 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: OK, I'm guessing this is a convention/practice I'm not familiar with. I did find the ordering of the functions baffling. If they're just in whatever order they were added, that makes sense of it.

Comment (CrazyIvan359) [reply Core/automation/lib/python/core/utils.py]: Maybe it's just me, but I think it makes the most sense to have them in the order added.

File: Core/automation/lib/python/core/utils.py (hunk 5 above)
Comment (CrazyIvan359) [reply Core/automation/lib/python/core/utils.py]: > I was trying to do similar to

The example you show does work the same, in fact only the name was changed in order to make everything snake_case and provide clearer function names.

File: Core/automation/lib/python/core/utils.py [hunk 8]
```diff
@@ -315,3 +316,61 @@ def postUpdate(item_or_item_name, new_value):
...
+                return state
+            elif return_type in type_methods:
+                if hasattr(state, type_methods[return_type]):
+                    state = getattr(state, type_methods[return_type])()
+            elif return_type in conv_methods:
+                return_type = conv_methods[return_type]
+            try: # attempt to cast
+                return return_type(state)
+            except:
+                LOG.warn(u"Item '{}' type '{}', cannot be converted to '{}'".format(item.name, type(state), return_type))
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py]: Just for clarity:

```python
LOG.warn(u"Item '{}' state type '{}', cannot be converted to '{}'".format(item.name, type(state), return_type))
```

Possibly confusing message though, `return_type` might be a conversion function not a type. Perhaps we should be using a new variable for the cast call.

File: Core/automation/lib/python/core/utils.py (hunk 7 above)
Comment (CrazyIvan359) [reply Core/automation/lib/python/core/utils.py]: I meant to put the link in actually, I agree it's a good reference to have!

> Much clearer! I think I'd forgotten the proper docstring format.

Practice makes perfect as they say, I wrote all the docs for MQTTany, Eos, and Selector among others.

File: Core/automation/lib/python/core/utils.py [hunk 9]
```diff
@@ -315,3 +316,61 @@ def postUpdate(item_or_item_name, new_value):
...
+                pass
+    else:
+        LOG.warn(u"Item does not exist or is uninitialized, returning default value")
```
Comment (CrazyIvan359) [inline Core/automation/lib/python/core/utils.py]: `validate_item` will give us a warning if the item doesn't exist. Giving a warning if the item is uninitialized is redundant, this is the expected behaviour of the function and the caller should be handling the return of the default value.

**Review by CrazyIvan359 — CHANGES_REQUESTED** [review]

File: Core/automation/lib/python/core/utils.py (hunk 9 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: Removed

File: Core/automation/lib/python/core/utils.py (hunk 8 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: How about
```python
LOG.warn(u"Item '{}' state is type '{}', cannot be converted to '{}'".format(item.name, type(state), return_type))
```

File: Core/automation/lib/python/core/utils.py (hunk 4 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py:40]: Kept due to `conv_methods` change

File: Core/automation/lib/python/core/utils.py (hunk 3 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py:23]: Kept due to `conv_methods` change

File: Core/automation/lib/python/core/utils.py (hunk 2 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: Removed

File: Core/automation/lib/python/core/utils.py (hunk 5 above)
Comment (politas) [reply Core/automation/lib/python/core/utils.py]: Removed

**Comment by CrazyIvan359** [comment]
Looks good, I'd like to do some testing before merging, and need to make sure it's compatible with OH2 as well (I don't see anything obvious that would make it incompatible). What tests have you done so far?

**Comment by politas** [comment]
I switched all my code over to using it, got half-way through re-writing them all to work with the changed behaviour, then decided to roll back to the working code using the deprecated functions and work through it all a bit more methodically. I've done a bunch of ad-hoc tests that were more focused on figuring out how it worked. Nothing really solid. What do you think about creating a test script that adds known items, uses them in known ways, then deletes the items, and reports when received results do not match expected results? Then it could be loaded up to any given OH instance to test different versions. It doesn't seem possible to run unit tests any way other than using OH, since the OH environment does not behave the same as the base Jython interpreter.

I'm happy to knock one up. I was thinking of placing it in `Core/automation/jsr223/python/personal/test-utils.py`? For this PR, I'll only include tests for `get_item_state()`, and it could then be expanded to provide unit testing for all the functions in further PRs.

Will be a few days though; I'm off to the coast for a short holiday.

**Comment by politas** [comment]
Testing suite is coming along well, apart from my rrd4j persistence DB trying to restore invalid values on my test item. @CrazyIvan359, do you know any way to turn off persistence via script? I see from the [doco](https://www.openhab.org/docs/configuration/persistence.html) that `item.persist()` forces persistence, and `item.persist("rrd4j")` will force an item to be persisted by an rrd4j database, but I'm not having any luck working out how to prevent persistence on an item. Neither `item.persist(None)` nor `item.persist("")` do it.

**Comment by CrazyIvan359** [comment]
There is a testing framework in `Core/automation/lib/python/core/testing.py` though I'm not very familiar with it. A simple script that creates items and tests the results using `assert` will work.

As for RRD4J I don't know, I'm running mapdb as my default/restore persistence service. Maybe just uninstall RRD4J?

**Comment by politas** [comment]
Yep, I'm using the testing framework.

**Comment by politas** [comment]
@CrazyIvan359 , I don't suppose you know how to create a `QuantityType` variable of a given unit? I'm trying to write a test for the unit conversion parameter, but for the life of me, I can't find anything on how that `QuantityType` class operates.

EDIT:
Ok, by trial and error, I have worked out that `test_value = QuantityType(u"55.0 °C")` does the trick.

**Comment by CrazyIvan359** [comment]
Looks like you beat me to it! I usually look at the OH code to find things like that, though the stubs should have helped and the *Reference* document in the stubs repo gives the locations of commonly used things like that.

I am working on getting a script working to generate stubs automatically, I will take a look at this when I get that working. The stubs I have are getting rather out of date and I'd like to be able to generate them as the builds happen.

**Comment by politas** [comment]
So the test suite as it is tests that most of the primary item (non-enumerated) types return the same value that is entered into an item, tests a unit conversion, and tests both types of return_type handling, though not exhaustively. Do you think we need exhaustive tests, or is testing each function adequate. I've also tested that None is returned for a non-existent item, and that the default is returned when one is given. I'll add another test for an uninitialised item, are there any other specific success or fail modes you think should be tested?

**Comment by rkoshak** [comment]
> do you know any way to turn off persistence via script?

No, there is no way to do that unless you happen to write a script to rewrite the rrd4j.persist file.

By default rrd4j will persist all supported Items on every change and every minute and do restoreOnStartup. If you want a different strategy (e.g. to control which Items get restoredOnStartup, limit which Items get saved in the first place) you'll have to create an rrd4j.persist file in conf/persist and define the strategy accordingly. 

When you call item.persist() what you are doing is overriding the default and saying "despite what the strategy is, persist this Item right now."

>  I'm running mapdb as my default/restore persistence service.

Note that if you set MapDB as the default none of the charting functionality built into OH 3 Pages will work. Your default persistence needs to be something that saves historic data beyond just the one value. You can still have MapDB of course and you can still have it restoreOnStartup. It just can't be the default. 

By default, OH 3 ships with rrd4j using the default strategy I described above.

**Comment by politas** [comment]
Just realised I'd forgotten to switch tabs to spaces in the test case file. I do think PEP's decision to standardise on spaces was a bad one, but we're following it, so...

Also added a test for uninitialised items.

**Comment by CrazyIvan359** [comment]
@politas
> Do you think we need exhaustive tests, or is testing each function adequate.

I think we should test just the most common conversions like:
* Number Item
  * `int`, `float`, maybe `BigDecimal`
* Number Item with units
  * `int`, `float`, maybe `BigDecimal`
* DateTime Item
  * each type we have a conversion function for
* other suggestions?

@rkoshak 
> Note that if you set MapDB as the default none of the charting functionality built into OH 3 Pages will work.

Now that you mention that I think I ran into that and changed the default to be Influx, but I still use MapDB for restore since it supports all item types (I found it an odd choice to make the default something that doesn't support all item types).

**Comment by politas** [comment]
Ok, If I run it using my modified run_tests with verbosity set to 2, I get these results now:
```
Starting test script
Running tests: 'CoreUtilsTestCase'
test_get_item_state_returns_ColorType_for_color_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_DateTimeType_for_date_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_DecimalType_for_number_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_PercentType_for_dimmer_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_StringType_for_string_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_ZonedDateTime_for_date_item_with_return_type (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_datetime_for_date_item_with_return_type (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_default_for_non_existing_item_with_default (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_color_for_color_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_datetime_for_datetime_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_float_for_number_item_with_return_type (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_int_for_number_item_with_return_type (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_java_calendar_for_DateTime_item_with_return_type (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_java_zonedatetime_for_DateTime_item_with_return_type (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_joda_datetime_for_DateTime_item_with_return_type (__main__.CoreUtilsTestCase) ... skipped 'Joda DateTime not supported'
test_get_item_state_returns_expected_number_for_number_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_percentage_for_dimmer_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_python_datetime_for_DateTime_item_with_return_type (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_string_for_number_item_with_return_type (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_expected_string_for_string_item (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_farenheit_for_number_item_with_unit (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_none_for_non_existing_item_with_no_default (__main__.CoreUtilsTestCase) ... ok
test_get_item_state_returns_none_for_uninitialized_item_with_no_default (__main__.CoreUtilsTestCase) ... ok

----------------------------------------------------------------------
Ran 23 tests in 10.288s

OK (skipped=1)
```
(I'll point out that the results seem to come in a random order. Presumably, unittest runs the different tests asynchronously.)

I don't think we're supporting BigDecimal?

**Comment by rkoshak** [comment]
> I found it an odd choice to make the default something that doesn't support all item types

Well, they needed something that was embeddable (which eliminates everything except MapDB, rrd4j, and SQLite), saves enough data to chart (which eliminates MapDB), and doesn't grow forever eventually eating all the disk space (which eliminates SQLite). rrd4j is the last DB standing. 

It supports all the data types that can be charted though. InfluxDB should be able to support all Item types though I've seen some reports that it's having trouble with Switches and Contacts.

**Comment by CrazyIvan359** [comment]
I was not aware of all of that. I have not had issues with switches, they appear as 0/1 now (I was doing this manually in OH2).

**Comment by politas** [comment]
Are there any outstanding changes I need to make?

**Comment by CrazyIvan359** [comment]
Nothing outstanding, I just haven't had time to do a thorough review and test it on OH2. I have another project that I need to get done first, I will get to this as soon as I can.

**Comment by politas** [comment]
@CrazyIvan359 how are you doing? Still planning to get back to this at some point in time?
